### PR TITLE
Ping check

### PR DESCRIPTION
**Pull request by jstanton617**
### What does this PR do?

Adds a ping (icmp echo) check, utilizing the system's `ping` command to avoid any permissions issues

### Motivation

Common support request

### Review checklist

- [x] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [x] Feature or bugfix has tests
- [x] Git history is clean
- [ ] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: ping/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+This check uses the system ping command, rather than generating the ICMP echo request
+itself, as creating an ICMP packet requires a raw socket, and creating raw sockets
+requires root privileges, which the agent does not have. The ping command uses the
+`setuid` access flag to run with elevated privileges, avoiding this issue.
+
+## Setup
+
+### Installation
+
+The Ping check is not included in the [Datadog Agent][2] package, so you will
```
Comment (ruthnaebeck) [inline ping/README.md]:
```suggestion
The Ping check is not included in the [Datadog Agent][2] package, so you
```

Comment (ruthnaebeck) [reply ping/README.md]: ^ Style guide, no future tense in docs.

File: ping/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+1. Edit the `ping.d/conf.yaml` file, in the `conf.d/` folder at the root of your
+   Agent's configuration directory to start collecting your ping performance data.
+   See the [sample ping.d/conf.yaml][2] for all available configuration options.
+
+2. [Restart the Agent][3]
+
+### Validation
+
+[Run the Agent's `status` subcommand][4] and look for `ping` under the Checks section.
```
Comment (ruthnaebeck) [inline ping/README.md]:
```suggestion
[Run the Agent's status subcommand][4] and look for `ping` under the Checks section.
```

File: ping/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+The Ping check is not included in the [Datadog Agent][2] package, so you will
+need to install it yourself.
+
+### Configuration
+
+1. Edit the `ping.d/conf.yaml` file, in the `conf.d/` folder at the root of your
+   Agent's configuration directory to start collecting your ping performance data.
+   See the [sample ping.d/conf.yaml][2] for all available configuration options.
+
+2. [Restart the Agent][3]
```
Comment (ruthnaebeck) [inline ping/README.md]:
```suggestion
2. [Restart the Agent][3].
```

File: ping/README.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+Returns DOWN if the check does not receive a response from `host`, otherwise UP.
+
+### Events
+
+Ping does not include any events.
+
+## Troubleshooting
+
+Need help? Contact [Datadog Support][5].
```
Comment (ruthnaebeck) [inline ping/README.md]:
```suggestion
Need help? Contact [Datadog support][5].
```

**Comment by ruthnaebeck** [comment]
If there is a logo for this integration, submit a PR here - https://github.com/DataDog/documentation/tree/master/src/images/integrations_logos

File: ping/README.md (hunk 1 above)
Comment (jstanton617) [reply ping/README.md]: Fixed

Looks like this needs to be fixed in the template as well: https://github.com/DataDog/integrations-core/blob/master/datadog_checks_dev/datadog_checks/dev/tooling/create.py#L59

**Comment by jstanton617** [comment]
> If there is a logo for this integration, submit a PR here - https://github.com/DataDog/documentation/tree/master/src/images/integrations_logos

https://github.com/DataDog/documentation/pull/3969

File: ping/tests/test_ping.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+def mock_exec_ping():
+    return """FAKEPING 127.0.0.1 (127.0.0.1): 56 data bytes
+64 bytes from 127.0.0.1: icmp_seq=0 ttl=64 time=0.093 ms
+
+--- 127.0.0.1 ping statistics ---
+1 packets transmitted, 1 packets received, 0.0% packet loss
+round-trip min/avg/max/stddev = 0.093/0.093/0.093/0.000 ms"""
+
+@mock.patch("datadog_checks.ping.PingCheck._exec_ping", return_value=mock_exec_ping())
+def test_check(aggregator, instance):
```
Comment (masci) [inline ping/tests/test_ping.py]: When you mock a function using the `@mock.patch` decorator, the resulting mock object is automagically passed as an arg to the decorated function, in this case `mock` thinks that `aggregator` is the mocked object. You can either add an argument to the function, with the caveat that also `pytest` uses function args in a magic way so you have to check what's the resulting order once mocked, or use `mock.patch` as a context manager (see next comment below) which seems easier here.

File: ping/tests/test_ping.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        c.check({'name': 'Datadog'})
+
+
+    test_check
+    # good check
+    instance = {
+        'host': '127.0.0.1',
+        'name': "Localhost"
+    }
+    c.check(instance)
```
Comment (masci) [inline ping/tests/test_ping.py]: An alternative approach to decorating the test function would be using `mock.patch` as a context manager within the function body. You have 2 alternatives:

1. mock the class **before** creating the instance, like you did by decorating the test function. This has the implication that every time you do `c = PingCheck('ping', {}, {})` **within** the `with` body, that specific `c` will result to be mocked.
2. mock the method **after** creating the instance, on the instance itself after you create it with `c = PingCheck('ping', {}, {})`

Option 2 might be more concise for this use case: 

```python
with mock.patch.object(c, "_exec_ping", return_value=mock_exec_ping()):
    c.check(instance)

aggregator.assert_service_check('network.ping.can_connect', AgentCheck.OK)
```
With the code above, the mock would be applied to the check **instance** and not the check class and the method will be mocked only within the `with` statement block, back to normal as soon as execution exits the block. So if you do `c.check({'name': 'Datadog'})` outside the block, `check` wouldn't be mocked and would behave as expected.

**Comment by miketheman** [comment]
Historical reference: https://github.com/DataDog/dd-agent/issues/599

**Pull request closed without merging**